scipy/special/_amos.c

```diff
@@ -0,0 +1,6206 @@
+        cy[i-1] = st*hci;
+        if ((st == 0.0) && (ey == 0.0)) { nz += 1; }
+    }
+    return nz;
+}
+
+
+int amos_binu(
+    double complex z,
+    double fnu,
+    int kode,
+    int n,
+    double complex *cy,
+    double rl,
+    double fnul,
+    double tol,
+    double elim,
+    double alim
+) {
+
+    //***BEGIN PROLOGUE  ZBINU
+    //***REFER TO  ZBESH,ZBESI,ZBESJ,ZBESK,ZAIRY,ZBIRY
+    //
+    //     ZBINU COMPUTES THE I FUNCTION IN THE RIGHT HALF Z PLANE
+    //
+    //***ROUTINES CALLED  AZABS,ZASYI,ZBUNI,ZMLRI,ZSERI,ZUOIK,ZWRSK
+    //***END PROLOGUE  ZBINU
+
+    double complex cw[2] = { 0. };
+    double az, dfnu;
+    int inw, nlast, nn, nui, nw, nz;
+
+    nz = 0;
+    az = cabs(z);
+    nn = n;
+    dfnu = fnu + n - 1;
+    if ((az <= 2.) || (az*az*0.25 <= (dfnu + 1.0))) {
+        /* GOTO 10 */
+        nw = amos_seri(z,fnu, kode, n, cy, tol, elim, alim);
+        inw = abs(nw);
+        nz += inw;
+        nn -= inw;
+        if (nn == 0) { return nz; }
+        if (nw >= 0) { return nz; }
+        dfnu = fnu + nn - 1;
+    }
+    /* GOTO 30 conditions*/
+    //
+    // ASYMPTOTIC EXPANSION FOR LARGE Z
+    //
+    if (az >= rl) {
+        if ((dfnu <= 1.0) || ((dfnu > 1.0) && (az+az >= dfnu*dfnu))) {
+            //
+            // MILLER ALGORITHM NORMALIZED BY THE SERIES
+            //
+            nw = amos_asyi(z, fnu, kode, n, cy, rl, tol, elim, alim);
+            if (nw < 0) {
+                nz = -1;
+                if (nw == -2) {
+                    nz = -2;
+                }
+                return nz;
+            }
+            return nz;
+        }
+    }
+    /* 40 */
+    if ((az < rl) || (az+az >= dfnu*dfnu)) {
+        nw = amos_mlri(z, fnu, kode, n, cy, tol);
+        if (nw < 0) {
+            nz = -1;
+            if (nw == -2) {
+                nz = -2;
+            }
+            return nz;
+        }
+        return nz;
+    }
+    /* 50 */
+    //
+    // OVERFLOW AND UNDERFLOW TEST ON I SEQUENCE FOR MILLER ALGORITHM
+    //
+    nw = amos_uoik(z, fnu, kode, 1, nn, cy, tol, elim, alim);
+    if (nw < 0) {
+        nz = -1;
+        if (nw == -2) { nz = -2; }
+        return nz;
+    }
+    nz += nw;
+    nn -= nw;
+    if (nn == 0) { return nz; }
+    dfnu = fnu + (nn -1);
+    /* GOTO 110s handled here */
+    if ((dfnu > fnul) || (az > fnul)) {
+        //
+        // INCREMENT FNU+NN-1 UP TO FNUL, COMPUTE AND RECUR BACKWARD
+        //
+        nui = (int)(fnul-dfnu) + 1;
+        nui = (nui > 0 ? nui : 0);
+        nw = amos_buni(z, fnu, kode, nn, cy, nui, &nlast, fnul, tol, elim, alim);
+        if (nw < 0) {
+            nz = -1;
+            if (nw == -2) { nz = -2; }
+            return nz;
+        }
+        nz += nw;
+        if (nlast == 0) { return nz; }
+        nn = nlast;
+    }
+    /* 60 */
+    if (az <= rl) {
+        /* 70 */
+        nw = amos_mlri(z, fnu, kode, n, cy, tol);
+        nz = -1;
+        if (nw == -2) { nz = -2; }
+        return nz;
```

Comment:
```suggestion
        if (nw < 0) {
            nz = -1;
            if (nw == -2) {
                nz = -2;
            }
            return nz;
        }
        return nz;
```

I've confirmed tests pass for me locally if this part is replaced with what follows the other call to `amos_mlri`.

Reply:
This still doesn't explain why scipy 1.11.x can't match 1e-15 rtol bound you showed though.

Reply:
This is what I get running the example above on `main`

```
In [2]: run test_iv_miller.py
scipy: (-2.0645908843310795e-55-1.112777373104039e-55j)
mpmath: (-2.06459088433096e-55 - 1.11277737310399e-55j)
relative error: 5.64730905928374e-14
```

so not quite as good as this branch now, but still better than this branch before the fix. I suspect you probably improved something somewhere else when translating, and this is where the improvement comes from verse `main`.